server/schedule/filter/region_filters.go

```diff
@@ -0,0 +1,207 @@
+		return statusRegionPendingPeer
+	}
+	return statusRegionOK
+}
+
+type regionDownFilter struct {
+	scope string
+}
+
+// NewRegionDownFilter creates a RegionFilter that filters all regions with down peers.
+func NewRegionDownFilter(scope string) RegionFilter {
+	return &regionDownFilter{scope: scope}
+}
+
+func (f *regionDownFilter) Scope() string {
+	return f.scope
+}
+
+func (f *regionDownFilter) Type() string {
+	return "Unhealthy"
+}
+
+func (f *regionDownFilter) Reason() string {
+	return "This region has too many down pees."
+}
+
+func (f *regionDownFilter) Select(region *core.RegionInfo) plan.Status {
+	if len(region.GetDownPeers()) > 0 {
+		return statusRegionDownPeer
+	}
+	return statusRegionOK
+}
+
+type regionReplicatedFilter struct {
+	scope   string
+	cluster Cluster
+}
+
+// NewRegionReplicatedFilter creates a RegionFilter that filters all unreplicated regions.
+func NewRegionReplicatedFilter(scope string, cluster Cluster) RegionFilter {
+	return &regionReplicatedFilter{scope: scope, cluster: cluster}
+}
+
+func (f *regionReplicatedFilter) Scope() string {
+	return f.scope
+}
+
+func (f *regionReplicatedFilter) Type() string {
+	return "NotReplicated"
+}
+
+func (f *regionReplicatedFilter) Reason() string {
+	if f.cluster.GetOpts().IsPlacementRulesEnabled() {
+		return "This region does not fit placement rule."
+	}
+	return "This region is not replicated"
+}
+
+func (f *regionReplicatedFilter) Select(region *core.RegionInfo) plan.Status {
+	if f.cluster.GetOpts().IsPlacementRulesEnabled() {
+		if !f.cluster.GetRuleManager().FitRegion(f.cluster, region).IsSatisfied() {
+			return statusRegionRule
+		}
+		return statusRegionOK
+	}
+	if !(len(region.GetLearners()) == 0 && len(region.GetPeers()) == f.cluster.GetOpts().GetMaxReplicas()) {
+		return statusRegionIsolation
+	}
+	return statusRegionOK
+}
+
+type regionEmptyFilter struct {
+	scope   string
+	cluster Cluster
+}
+
+// NewRegionEmptyFilter returns creates a RegionFilter that filters all empty regions.
+func NewRegionEmptyFilter(scope string, cluster Cluster) RegionFilter {
+	return &regionEmptyFilter{scope: scope, cluster: cluster}
+}
+
+func (f *regionEmptyFilter) Scope() string {
+	return f.scope
+}
+
+func (f *regionEmptyFilter) Type() string {
+	return "EmptyRegion"
+}
+
+func (f *regionEmptyFilter) Reason() string {
+	return "This region is empty"
+}
+
+func (f *regionEmptyFilter) Select(region *core.RegionInfo) plan.Status {
+	if !isEmptyRegionAllowBalance(f.cluster, region) {
+		return statusRegionEmpty
+	}
+	return statusRegionOK
+}
+
+// isEmptyRegionAllowBalance checks if a region is an empty region and can be balanced.
+func isEmptyRegionAllowBalance(cluster Cluster, region *core.RegionInfo) bool {
+	return region.GetApproximateSize() > core.EmptyRegionApproximateSize || cluster.GetRegionCount() < core.InitClusterRegionThreshold
+}
+
+// Cluster provides an overview of a cluster's regions distribution.
+type Cluster interface {
```

Comment:
Can we avoid defining this interface?

Reply:
I defined this because of avoiding import cycles

Reply:
defining this interface looks ok to me, However could we make the interface private?